grudge/dt_utils.py

```diff
@@ -271,25 +271,33 @@ def dt_geometric_factors(
     dd_face = DOFDesc("all_faces", dd.discretization_tag)
     face_discr = dcoll.discr_from_dd(dd_face)
 
-    # To get a single value for the total surface area of a cell, we
-    # take the sum over the averaged face areas on each face.
-    # NOTE: The face areas are the *same* at each face nodal location.
-    # This assumes there are the *same* number of face nodes on each face.
     surface_areas = abs(
         op.elementwise_integral(
             dcoll, dd_face, face_discr.zeros(actx) + 1.0
         )
     )
+
+    # Sum the areas of each face to get the total surface area
     surface_areas = DOFArray(
         actx,
         data=tuple(
-            actx.einsum("fej->e",
-                        face_ae_i.reshape(
-                            vgrp.mesh_el_group.nfaces,
-                            vgrp.nelements,
-                            afgrp.nunit_dofs
-                        ),
-                        tagged=(FirstAxisIsElementsTag(),)) / afgrp.nunit_dofs
+            actx.einsum(
+                "fej->e",
+                face_ae_i.reshape(
+                    vgrp.mesh_el_group.nfaces,
+                    vgrp.nelements,
+                    -1
+                ),
+                tagged=(FirstAxisIsElementsTag(),)) / (
+                    # NOTE: Whenever the array context can't perform nonscalar
+                    # broadcasting, elementwise reductions
+                    # (like `elementwise_integral`) repeat the scalar value of
+                    # the reduction at each degree of freedom. To get a single
+                    # value for the total surface area of a cell,
+                    # we sum over all nodes and average.
+                    1 if actx.supports_nonscalar_broadcasting
+                    else afgrp.nunit_dofs
```

Comment:
This nodal averaging is super sketchy to me. Why not just use honest quadrature (using `afgrp.weights` or some such) to compute areas?

Reply:
@inducer I agree this is a bit ugly, but let me explain why this was done.

You recall the issues with nonscalar broadcasting for array contexts right? Well when an array context can't support it, you get repeated values at each node, rather than the single `DOFArray` with one value per cell. This tried to address this difference.

> Why not just use honest quadrature (using afgrp.weights or some such) to compute areas?

Um, that's what is happening if you look up a few more lines. In the definition of `surface_areas`: https://github.com/inducer/grudge/blob/be1e72405c4bdb9ec7dc27462bfd77001b96c193/grudge/dt_utils.py#L274

Reply:
Sounds good! Yeah I like that, I'll update accordingly

Reply:
Updated: https://github.com/inducer/grudge/pull/195/commits/4f04de2e48b829aca1050e298aaf5d6d7a75237e